### PR TITLE
Implement `SymbolicOp2` and `Adjoint2`

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**

Add a `SymbolicOp2` and `Adjoint2` that inherit from `Operator2`.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-120432]
[sc-121615]

### CODE REVIEW

File: pennylane/ops/op_math/adjoint2.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    @classproperty
+    @classmethod
+    @override
+    def has_adjoint(cls):
+        return True
+
+    @override
+    def adjoint(self):
+        return type(self.base)(**self.base.arguments)
```
Comment (astralcai) [inline pennylane/ops/op_math/adjoint2.py:93]: Is this how you do it now?

File: pennylane/ops/op_math/adjoint2.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the base class for the adjoint of operators."""
+
+from typing_extensions import override
+
+import pennylane as qp
```
Comment (andrijapau) [inline pennylane/ops/op_math/adjoint2.py:19]: Can we clean up this import as best we can? Seems like it's only used for PauliSentence - can we just do a local import under `pauli_rep`? 🤔

Comment (astralcai) [reply pennylane/ops/op_math/adjoint2.py:19]: Is local imports the recommended solution for these scenarios? I'm not sure if it is actually preferred.

Comment (andrijapau) [reply pennylane/ops/op_math/adjoint2.py:19]: @albi3ro

Comment (andrijapau) [reply pennylane/ops/op_math/adjoint2.py:19]: I just don't see why it should be a global import if it's used in a single method 🤔

File: pennylane/ops/op_math/adjoint.py [hunk 3]
```diff
@@ -180,9 +181,9 @@ def loop_fn(i):
 def create_adjoint_op(fn, lazy):
     """Main logic for qp.adjoint, but allows bypassing the compiler dispatch if needed."""
     if qp.math.is_abstract(fn):
-        return Adjoint(fn)
-    if isinstance(fn, Operator):
-        return Adjoint(fn) if lazy else _single_op_eager(fn, update_queue=True)
+        return _make_adjoint_op(fn)
+    if isinstance(fn, (Operator, Operator2)):
```
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint.py]: Won't need this after #9596

File: pennylane/ops/op_math/adjoint2.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    @classproperty
+    @classmethod
+    @override
+    def has_adjoint(cls):
+        return True
+
+    @override
+    def adjoint(self):
+        return type(self.base)(**self.base.arguments)
```
Comment (andrijapau) [inline pennylane/ops/op_math/adjoint2.py:93]: What's the motivation for rebuilding it here? Is this basically the same as flatten unflatten?

File: pennylane/ops/op_math/adjoint2.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    @override
+    def simplify(self):
+        base = self.base.simplify()
+        if base.has_adjoint:
+            return base.adjoint().simplify()
+        return Adjoint2(base)
```
Comment (andrijapau) [inline pennylane/ops/op_math/adjoint2.py:100]: When comparing with the legacy `Adjoint`, seems like we dropped the capture guard, is this intentional?

Old ⤵️ 
```python
    def simplify(self):
        base = self.base if qp.capture.enabled() else self.base.simplify()
        if base.has_adjoint:
            return base.adjoint() if qp.capture.enabled() else base.adjoint().simplify()
        return Adjoint(base=base)
```

File: pennylane/ops/op_math/adjoint2.py (hunk 4 above)
Comment (astralcai) [reply pennylane/ops/op_math/adjoint2.py:93]: Yep... That's the idea.

File: pennylane/ops/op_math/adjoint2.py (hunk 5 above)
Comment (astralcai) [reply pennylane/ops/op_math/adjoint2.py:100]: Yes, I asked @albi3ro what they were for and she told me that it was most likely a patch for something that made sense at one point. Since operators are going to be different completely, we go with the simplest implementation here for now, and if something does break, we can come back to it.

Comment (andrijapau) [reply pennylane/ops/op_math/adjoint2.py:100]: Ok, got it

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -463,6 +463,9 @@
   [(#9607)](https://github.com/PennyLaneAI/pennylane/pull/9607)
   [(#9627)](https://github.com/PennyLaneAI/pennylane/pull/9627)
 
+* Added ``SymbolicOp2``, ``Adjoint2`` that inherit from the new ``Operator2``.
+  [(#9597)](https://github.com/PennyLaneAI/pennylane/pull/9597)
+
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md]: Can you put this under the same entry as the other `Operator2` PRs?

File: pennylane/ops/op_math/adjoint2.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/ops/op_math/adjoint2.py:19]: https://peps.python.org/pep-0008/#imports

Imports should always be top-level when possible

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (astralcai) [reply doc/releases/changelog-dev.md]: [4412aa3](https://github.com/PennyLaneAI/pennylane/pull/9597/commits/4412aa3ab04d44fe0b1623d60a6fe0354ffef648)

File: pennylane/ops/op_math/adjoint.py (hunk 3 above)
Comment (astralcai) [reply pennylane/ops/op_math/adjoint.py]: [a55a410](https://github.com/PennyLaneAI/pennylane/pull/9597/commits/a55a41097c3c3803cc9ecda88ed534b7578d9259)

File: pennylane/ops/op_math/adjoint.py [hunk 7]
```diff
@@ -493,4 +493,12 @@ def generator(self):
...
 
 
+def _make_adjoint_op(op: Operator) -> Adjoint | Adjoint2:
+    return Adjoint2(op) if isinstance(op, Operator2) else Adjoint(op)
+
+
+def _is_operator(op):
+    return isinstance(op, Operator)
+
+
```
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint.py]: do we still need this?

Comment (astralcai) [reply pennylane/ops/op_math/adjoint.py]: good point lol

Comment (astralcai) [reply pennylane/ops/op_math/adjoint.py]: [463a7bd](https://github.com/PennyLaneAI/pennylane/pull/9597/commits/463a7bd77ef293b2ff5108f0eaa0c9a3a0c8d3f1)

File: pennylane/ops/op_math/adjoint2.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    def __init__(self, base: Operator2):
+        super().__init__(base)
+
+    @property
+    @override
+    def pauli_rep(self):
+        if not self.base.pauli_rep:
+            return None
+        rep = {pw: math.conjugate(c) for pw, c in self.base.pauli_rep.items()}
+        return qp.pauli.PauliSentence(rep)
```
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint2.py]: Should we bother caching this?

File: pennylane/ops/op_math/adjoint2.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    @staticmethod
+    @override
+    def compute_sparse_matrix(base, format="csr"):  # pylint: disable=arguments-differ
+        base_matrix = base.sparse_matrix()
+        return math.transpose(math.conj(base_matrix)).asformat(format=format)
+
+    @staticmethod
+    @override
+    def compute_eigvals(base):  # pylint: disable=arguments-differ
+        return math.conj(base.eigvals())
```
Comment (albi3ro) [inline pennylane/ops/op_math/adjoint2.py:73]: ❤️

File: tests/devices/qubit/test_jaxpr_adjoint_jvp.py [hunk 10]
```diff
@@ -92,7 +93,7 @@ def f(x):
 
         jaxpr = jax.make_jaxpr(f)(0.5).jaxpr
 
-        with pytest.raises(ValueError):
+        with pytest.raises(GeneratorUndefinedError):
```
Comment (albi3ro) [inline tests/devices/qubit/test_jaxpr_adjoint_jvp.py:96]: That is actually a much better error message 👍

File: pennylane/ops/op_math/adjoint2.py (hunk 8 above)
Comment (astralcai) [reply pennylane/ops/op_math/adjoint2.py]: Might as well. [2de5a9e](https://github.com/PennyLaneAI/pennylane/pull/9597/commits/2de5a9e819697d1010e3853b5e94be59c454c4e0)

File: tests/gradients/parameter_shift/test_parameter_frequencies.py [hunk 11]
```diff
@@ -34,7 +34,6 @@ def test_parameter_frequencies_raises_error_too_many_dynamic_args(self):
         """Test that parameter_frequencies raises an error if there are too many dynamic arguments."""
 
         class MultiArgOpWithGen(Operator2):
-            num_params = 2
             num_wires = 1
```
Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_frequencies.py]: Potentially out of scope, but mind getting rid of `num_wires` here and in the other operators too?

Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_frequencies.py]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_frequencies.py [hunk 12]
```diff
@@ -54,7 +53,6 @@ def test_parameter_frequencies_raises_error_no_generator(self):
         """Test that parameter_frequencies raises an error if the op.generator() is undefined."""
 
         class SingleArgOpNoGen(Operator2):
-            num_params = 1
             num_wires = 1
```
Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_frequencies.py]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_frequencies.py [hunk 13]
```diff
@@ -73,7 +71,6 @@ def test_param_freqs_no_generator(self, freqs):
         """Test that parameter_frequencies are accessible when provided explicitly, even if the op.generator() is undefined."""
 
         class MultiArgOpNoGenParamFreqs(Operator2):
-            num_params = 2
             num_wires = 1
```
Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_frequencies.py]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_frequencies.py [hunk 14]
```diff
@@ -102,14 +99,10 @@ def test_param_freqs_with_generator(self, matrix):
         """Test that parameter_frequencies relate to the eigenvalues of the generator if the op.generator() is defined."""
 
         class OpWithGen(Operator2):
-            num_params = 1
             num_wires = 1
```
Comment (albi3ro) [inline tests/gradients/parameter_shift/test_parameter_frequencies.py]:
```suggestion
```

File: tests/gradients/parameter_shift/test_parameter_frequencies.py (hunk 11 above)
Comment (astralcai) [reply tests/gradients/parameter_shift/test_parameter_frequencies.py]: [d5d13cb](https://github.com/PennyLaneAI/pennylane/pull/9597/commits/d5d13cbf7853d6ea2e8fefb46a1fa976cbffffaf)

**Review by albi3ro — APPROVED** [review]

**Review by andrijapau — APPROVED** [review]
Changes look good to me 🙂 :shipit:
